File: blog-web-app/python/tests/blog_builder.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timedelta, timezone

from blog_web_app import Blog, Post

from .fixed_clock import FixedClock


@dataclass
class _SeededPost:
    user: str
    title: str
    body: str
    minutes_after_start: int


@dataclass
class _SeededComment:
    post_index: int
    user: str
    body: str
    minutes_after_start: int


@dataclass
class _SeededTag:
    post_index: int
    tag: str


class BlogBuilder:
    def __init__(self) -> None:
        self._start_time = datetime(2026, 1, 15, 9, 0, 0, tzinfo=timezone.utc)
        self._seeded_posts: list[_SeededPost] = []
        self._seeded_comments: list[_SeededComment] = []
        self._seeded_tags: list[_SeededTag] = []

    def starting_at(self, dt: datetime) -> BlogBuilder:
        self._start_time = dt
        return self

    def with_post(
        self, user: str, title: str, body: str, minutes_after_start: int = 0
    ) -> BlogBuilder:
        self._seeded_posts.append(_SeededPost(user, title, body, minutes_after_start))
        return self

    def with_comment(
        self, post_index: int, user: str, body: str, minutes_after_start: int = 0
    ) -> BlogBuilder:
        self._seeded_comments.append(
            _SeededComment(post_index, user, body, minutes_after_start)
        )
        return self

    def with_tag(self, post_index: int, tag: str) -> BlogBuilder:
        self._seeded_tags.append(_SeededTag(post_index, tag))
        return self

    def build(self) -> tuple[Blog, FixedClock, list[Post]]:
        clock = FixedClock(self._start_time)
        blog = Blog(clock)
        posts: list[Post] = []

        for seed in self._seeded_posts:
            clock.advance_to(
                self._start_time + timedelta(minutes=seed.minutes_after_start)
            )
            posts.append(blog.create_post(seed.user, seed.title, seed.body))

        for seed in self._seeded_comments:
            clock.advance_to(
                self._start_time + timedelta(minutes=seed.minutes_after_start)
            )
            blog.add_comment(seed.user, posts[seed.post_index].id, seed.body)

        for seed in self._seeded_tags:
            blog.add_tag(
                posts[seed.post_index].author,
                posts[seed.post_index].id,
                seed.tag,
            )

        return blog, clock, posts

File: blog-web-app/typescript/tests/BlogBuilder.ts
import { Blog } from '../src/Blog.js';
import { Post } from '../src/Post.js';
import { FixedClock } from './FixedClock.js';

interface SeededPost {
  user: string;
  title: string;
  body: string;
  minutesAfterStart: number;
}

interface SeededComment {
  postIndex: number;
  user: string;
  body: string;
  minutesAfterStart: number;
}

interface SeededTag {
  postIndex: number;
  tag: string;
}

export class BlogBuilder {
  private _startTime = new Date(Date.UTC(2026, 0, 15, 9, 0, 0));
  private _seededPosts: SeededPost[] = [];
  private _seededComments: SeededComment[] = [];
  private _seededTags: SeededTag[] = [];

  startingAt(date: Date): this {
    this._startTime = date;
    return this;
  }

  withPost(user: string, title: string, body: string, minutesAfterStart = 0): this {
    this._seededPosts.push({ user, title, body, minutesAfterStart });
    return this;
  }

  withComment(postIndex: number, user: string, body: string, minutesAfterStart = 0): this {
    this._seededComments.push({ postIndex, user, body, minutesAfterStart });
    return this;
  }

  withTag(postIndex: number, tag: string): this {
    this._seededTags.push({ postIndex, tag });
    return this;
  }

  build(): { blog: Blog; clock: FixedClock; posts: Post[] } {
    const clock = new FixedClock(this._startTime);
    const blog = new Blog(clock);
    const posts: Post[] = [];

    for (const { user, title, body, minutesAfterStart } of this._seededPosts) {
      clock.advanceTo(new Date(this._startTime.getTime() + minutesAfterStart * 60_000));
      posts.push(blog.createPost(user, title, body));
    }

    for (const { postIndex, user, body, minutesAfterStart } of this._seededComments) {
      clock.advanceTo(new Date(this._startTime.getTime() + minutesAfterStart * 60_000));
      blog.addComment(user, posts[postIndex]!.id, body);
    }

    for (const { postIndex, tag } of this._seededTags) {
      blog.addTag(posts[postIndex]!.author, posts[postIndex]!.id, tag);
    }

    return { blog, clock, posts };
  }
}
